### PR TITLE
cl/merkle_tree: add EIP-7495 ProgressiveContainer root

### PR DESCRIPTION
**Pull request by JackCC703**
## Summary

Add `ProgressiveContainerRoot` following EIP-7495.

The helper:

- places active field roots at their stable indices
- zero-fills inactive positions
- combines `MerkleizeProgressive` with `MixInActiveFields`
- rejects invalid `active_fields` configurations

Tests cover the EIP-7807 18-field layout, sparse fields, the 256-bit boundary, invalid configurations, and input immutability.

## References

- EIP-7495: `ethereum/EIPs@c81d843b3f8aa839fe42911c5b6e501c7d2940a3`
- Reference implementation: `ethereum/remerkleable@2f0baeef0082d4278acaef7d822deb7009d7db7e`
- EIP-7807: `ethereum/EIPs@75d7bc2c20a91ec017d147f400d6bbf767843e2c`

Reference vectors were generated from the pinned `remerkleable` revision.

## Testing

- `go test ./cl/merkle_tree -run '^TestProgressiveContainerRoot' -count=1`
- `go test ./cl/merkle_tree -count=1`
- `go test -race ./cl/merkle_tree -count=1`
- `go tool golangci-lint run --config ./.golangci.yml ./cl/merkle_tree/...`
- `make erigon integration`

Depends on #22528.

### CODE REVIEW

File: cl/merkle_tree/progressive.go [hunk 1]
```diff
@@ -41,6 +41,65 @@ func ProgressiveListRoot(chunks [][32]byte, logicalLength uint64) ([32]byte, err
 	return crypto.Sha256(progressiveRoot[:], lengthRoot[:]), nil
 }
 
+// MixInActiveFields computes the EIP-7495 active-fields mix-in. Bit i is
+// packed into bit i%8 of byte i/8 in a zero-padded 32-byte chunk.
+func MixInActiveFields(root [32]byte, activeFields []bool) ([32]byte, error) {
+	if len(activeFields) > 256 {
+		return [32]byte{}, errors.New("active fields exceed 256 bits")
+	}
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go:49]: `MixInActiveFields` is currently only checking `len(activeFields) > 256`. But as per EIP-7495 spec, an `active_fields` configuration cannot be empty (`len == 0`) and cannot end with a `0` bit (`!activeFields[len-1]`).

Right now `ProgressiveContainerRoot` validates these edge cases, but since `MixInActiveFields` is exported publicly, if someone calls it directly with something like `[]bool{true, false}` or empty slice, it silently hashes without giving error.

Can we add these checks here as well?

```go
if len(activeFields) == 0 {
    return [32]byte{}, errors.New("active fields cannot be empty")
}
if !activeFields[len(activeFields)-1] {
    return [32]byte{}, errors.New("active fields must end with an active field")
}
```

File: cl/merkle_tree/progressive.go [hunk 2]
```diff
@@ -41,6 +41,65 @@ func ProgressiveListRoot(chunks [][32]byte, logicalLength uint64) ([32]byte, err
...
+	for _, active := range activeFields {
+		if active {
+			activeFieldCount++
+		}
+	}
+	if activeFieldCount != len(fieldRoots) {
+		return [32]byte{}, errors.New("active field count does not match field roots")
+	}
+
+	expandedRoots := make([][32]byte, len(activeFields))
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go:87]: here `make([][32]byte, len(activeFields))` will allocate on heap for every call.

In actual execution / consensus paths, almost all progressive containers have `≤ 32` fields (for example `ExecutionPayload` in EIP-7807 has only 18 fields).

We can do one small optimization here: use a small stack array `var stackRoots [32][32]byte` for `len(activeFields) <= 32` so we completely avoid heap allocations on hot paths (similar to `maxStackLeaves = 32` in `merkle_root.go`).

Something like:
```go
var stackRoots [32][32]byte
var expandedRoots [][32]byte
if len(activeFields) <= 32 {
    expandedRoots = stackRoots[:len(activeFields)]
} else {
    expandedRoots = make([][32]byte, len(activeFields))
}
```
This will save GC pressure when processing blocks repeatedly.

File: cl/merkle_tree/progressive.go [hunk 3]
```diff
@@ -41,6 +41,65 @@ func ProgressiveListRoot(chunks [][32]byte, logicalLength uint64) ([32]byte, err
...
+
+	activeFieldCount := 0
+	for _, active := range activeFields {
+		if active {
+			activeFieldCount++
+		}
+	}
+	if activeFieldCount != len(fieldRoots) {
+		return [32]byte{}, errors.New("active field count does not match field roots")
+	}
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go:85]: Also here, currently we are iterating over `activeFields` 3 times in total (first time to count 1s, second time to populate `expandedRoots`, and third time inside `MixInActiveFields`).

We can actually merge this count validation directly into the `expandedRoots` loop in a single pass:

```go
fieldIndex := 0
for i, active := range activeFields {
    if active {
        if fieldIndex >= len(fieldRoots) {
            return [32]byte{}, errors.New("active field count does not match field roots")
        }
        expandedRoots[i] = fieldRoots[fieldIndex]
        fieldIndex++
    }
}
if fieldIndex != len(fieldRoots) {
    return [32]byte{}, errors.New("active field count does not match field roots")
}
```
That way loop pass is reduced and it also fails fast if `fieldRoots` count is less than active bits.

File: cl/merkle_tree/progressive_test.go [hunk 4]
```diff
@@ -93,6 +94,210 @@ func TestProgressiveListRootReferenceVectors(t *testing.T) {
 	}
 }
 
+func TestMixInActiveFieldsReferenceVectors(t *testing.T) {
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive_test.go:97]: here, all current reference vector tests (sparseFields, boundaryFields, etc.), field index 0 is always set to true.

Could we add a small unit test where field index 0 is inactive (`activeFields = []bool{false, true, true}`)? Just to ensure zero-filling at leaf position 0 / gindex 4 works as expected when the first field itself is absent.

File: cl/merkle_tree/progressive_test.go [hunk 5]
```diff
@@ -93,6 +94,210 @@ func TestProgressiveListRootReferenceVectors(t *testing.T) {
...
+			expected:     "active fields exceed 256 bits",
+		},
+		{
+			name:         "trailing inactive field",
+			fieldRoots:   progressiveTestChunks(1),
+			activeFields: []bool{true, false},
+			expected:     "active fields must end with an active field",
+		},
+		{
+			name:         "active field count mismatch",
```
Comment (domiwei) [inline cl/merkle_tree/progressive_test.go]: Could we also cover the reverse count mismatch here—fewer field roots than active bits, for example one root with `[]bool{true, true}`? The equality check handles it today, but an explicit test would lock down the symmetric error path and ensure it returns the expected error without indexing past `fieldRoots`.

File: cl/merkle_tree/progressive.go (hunk 1 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive.go]: Thanks, added the empty and trailing-inactive checks to `MixInActiveFields`. 
Both it and `ProgressiveContainerRoot` now share `validateActiveFields`, and I added tests for the invalid direct-call cases.

File: cl/merkle_tree/progressive.go (hunk 2 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive.go]: Updated this to reuse `maxStackLeaves`: configurations with up to 32 positions use a stack-backed array, while larger configurations fall back to `make`.

File: cl/merkle_tree/progressive.go (hunk 3 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive.go]: I merged active-bit counting and root placement into one pass, with a bounds check for too few field roots and a final check for too many field roots.

File: cl/merkle_tree/progressive_test.go (hunk 4 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive_test.go]: Added a reference-vector case with `activeFields = []bool{false, true, true}` to verify zero-filling at stable field position 0.

File: cl/merkle_tree/progressive_test.go (hunk 5 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive_test.go]: Added the symmetric mismatch case with one field root and `[]bool{true, true}`. 
It now verifies that the function returns the expected error without indexing past `fieldRoots`.

File: cl/merkle_tree/progressive.go [hunk 6]
```diff
@@ -41,6 +41,75 @@ func ProgressiveListRoot(chunks [][32]byte, logicalLength uint64) ([32]byte, err
...
+
+	var packed [32]byte
+	for i, active := range activeFields {
+		if active {
+			packed[i/8] |= 1 << (uint(i) % 8)
+		}
+	}
+
+	return crypto.Sha256(root[:], packed[:]), nil
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go]: one micro-optimization nit here. Currently `crypto.Sha256(left[:], right[:])` passes slices variadically `(...[]byte)`, which causes Go runtime to allocate a small 64-byte slice header on the heap for concatenation.

If we use a stack-allocated 64-byte array (similar to `hashPair` in `merkle_root.go`), we can make this 100% allocation-free:

```go
var buf [64]byte
copy(buf[:32], left[:])
copy(buf[32:], right[:])
return sha256.Sum256(buf[:]), nil
```
Micro-nit, but avoids slice allocations during tree node hashing.

File: cl/merkle_tree/progressive.go
Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go:76]: same here [as](https://github.com/erigontech/erigon/pull/23130/changes#r3788437933)

Comment (Sahil-4555) [inline cl/merkle_tree/progressive.go:61]: Even though optimized `expandedRoots` to be stack-allocated in `ProgressiveContainerRoot`, inside `merkleizeProgressive` it still allocates `subtree` on the heap at every recursive step because `MerkleizeVector` modifies input in-place. Basically as an idea, we can pass a small stack scratch buffer (like `[64][32]byte`) down `merkleizeProgressive` and `copy` the subtree chunks into that scratch space instead of calling `append([][32]byte(nil), ...)`. That way the entire recursive progressive tree construction becomes 100% zero-allocation for  ≤ 32 leaves.

File: cl/merkle_tree/progressive_test.go [hunk 7]
```diff
@@ -93,6 +94,247 @@ func TestProgressiveListRootReferenceVectors(t *testing.T) {
...
+	}
+
+	sparseFields := make([]bool, 18)
+	for _, i := range []int{0, 7, 8, 17} {
+		sparseFields[i] = true
+	}
+
+	boundaryFields := make([]bool, 256)
+	boundaryFields[0] = true
+	boundaryFields[255] = true
```
Comment (Sahil-4555) [inline cl/merkle_tree/progressive_test.go:140]: Currently `boundaryFields` has both index 0 and index 255 set to `true`. Could we also add a test case where fields $0 \ldots 254$ are all `false` and only field index 255 is `true` (`onlyLastField255Active := make([]bool, 256); onlyLastField255Active[255] = true`)?

This will test zero-chunk filling across all 255 initial positions up to the 256-bit boundary

File: cl/merkle_tree/progressive.go (hunk 6 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive.go]: Thanks. I checked the current `common/crypto.Sha256` implementation: inputs totaling up to 64 bytes already use a stack-allocated `[64]byte` buffer, and `TestSha256AllocFree` verifies that the 32+32-byte path has zero allocations. I’ll keep the shared helper here to avoid duplicating the same fast path.

File: cl/merkle_tree/progressive.go
Comment (JackCC703) [reply cl/merkle_tree/progressive.go:76]: Same here: both inputs are 32 bytes, so this call also uses the existing allocation-free 64-byte stack path in `common/crypto.Sha256`.

Comment (JackCC703) [reply cl/merkle_tree/progressive.go:61]: Good point—the subtree copy currently allocates once per non-empty progressive level. Since `MerkleizeProgressive` is shared existing code and a fixed scratch buffer only covers the small-input path, I’d prefer to handle this as a separate benchmark-backed optimization with a correct fallback for larger progressive inputs.

File: cl/merkle_tree/progressive_test.go (hunk 7 above)
Comment (JackCC703) [reply cl/merkle_tree/progressive_test.go:140]: Thanks, I added a ProgressiveContainerRoot reference-vector case with only field index 255 active, covering zero-filling across the preceding 255 positions.

**Comment by JackCC703** [comment]
Hi @Sahil-4555 @domiwei, I noticed that #22912 has added similar progressive container helpers to `main`, so this PR now overlaps with it.

Would you prefer me to close this PR as superseded, or narrow it to any missing parts, such as `MixInActiveFields` or additional tests?

After that, I plan to move on to the EIP-7807 execution payload adapter and proof layer.
